### PR TITLE
【Hackathon No.2】

### PR DESCRIPTION
**Pull request by yangguohao**
### PR types
New Features

### PR changes
OPs

### Description
cdist op
rfc:https://github.com/PaddlePaddle/community/pull/538

### CODE REVIEW

**Comment by yangguohao** [comment]
您好，我尝试先用 python 写了一个 cdist 的算法。我想请问一下在 CUDA/C++ 中有没有 concat 和 repeat_interleave 类似的api。

**Comment by GGBond8488** [comment]
> 您好，我尝试先用 python 写了一个 cdist 的算法。我想请问一下在 CUDA/C++ 中有没有 concat 和 repeat_interleave 类似的api。

Hello, paddle的 c++端有对应的op以及kernel的，但是更建议直接用组合的方式在python端实现对应的api

**Comment by yangguohao** [comment]
> > 您好，我尝试先用 python 写了一个 cdist 的算法。我想请问一下在 CUDA/C++ 中有没有 concat 和 repeat_interleave 类似的api。
> 
> Hello, paddle的 c++端有对应的op以及kernel的，但是更建议直接用组合的方式在python端实现对应的api

您好，感谢回复。`更建议直接用组合的方式在python端实现` 这句话我理解的意思是可以直接用 paddle.xx 的形式来实现（也就是我目前实现的方法，只是缺少 inf 和 0 的情况），是这样吗？

**Comment by GGBond8488** [comment]
> > > 您好，我尝试先用 python 写了一个 cdist 的算法。我想请问一下在 CUDA/C++ 中有没有 concat 和 repeat_interleave 类似的api。
> > 
> > 
> > Hello, paddle的 c++端有对应的op以及kernel的，但是更建议直接用组合的方式在python端实现对应的api
> 
> 您好，感谢回复。`更建议直接用组合的方式在python端实现` 这句话我理解的意思是可以直接用 paddle.xx 的形式来实现（也就是我目前实现的方法，只是缺少 inf 和 0 的情况），是这样吗？

对的，需要增加检验，处理边界值等逻辑, 在pr Description 附上rfc设计文档的链接吧

**Comment by yangguohao** [comment]
@GGBond8488 您好，增加了检验，处理边界条件等内容，麻烦再 review 下

File: python/paddle/tensor/linalg.py [hunk 1]
```diff
@@ -3350,3 +3350,58 @@ def corrcoef(x, rowvar=True, name=None):
...
+    x = paddle.to_tensor([[1, 3],[3, 4]], dtype="float32")
+    y = paddle.to_tensor([[3, 2],[3, 1]], dtype="float32")
+
+    out = paddle.cdist(x, y, 2)
+    print(out) # out = [[2.23606801, 2.82842708],
+               #        [2.        , 3.        ]])
+
+    """
+    check_variable_and_dtype(x, 'dtype', ['float32', 'float64'], 'cdist')
+    check_variable_and_dtype(y, 'dtype', ['float32', 'float64'], 'cdist')
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py:3395]: p也加一个限制吧

File: python/paddle/tensor/linalg.py [hunk 2]
```diff
@@ -3350,3 +3350,58 @@ def corrcoef(x, rowvar=True, name=None):
...
+    x_shape = x.shape
+    y_shape = y.shape
+    if len(x_shape) < 2 or len(y_shape) < 2:
+        raise ValueError('The Input(x) and Input(y) should be 2-D tensor.')
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py]: should be 2-D tensor不对吧，应该至少是2D tensor

File: python/paddle/tensor/linalg.py [hunk 3]
```diff
@@ -3350,3 +3350,58 @@ def corrcoef(x, rowvar=True, name=None):
...
+    if x_shape[:-2] != y_shape[:-2] or x_shape[-1] != y_shape[-1]:
+        raise ValueError(
+            "Input(x) shape[:-2] should be same as Input(y) shape[:-2] "
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py]: Input(x) shape[:-2] should be same as Input(y) shape[:-2] 确定需要相等吗，根据定义应该是可以不相等吧

File: python/paddle/tensor/linalg.py [hunk 4]
```diff
@@ -3350,3 +3350,58 @@ def corrcoef(x, rowvar=True, name=None):
...
+            "and Input(x) shape[-1] shoule be same as Input(y) shape[-1]"
+        )
+    y = paddle.concat([y] * x_shape[-2], axis=-2)
+    x = paddle.repeat_interleave(x, y_shape[-2], axis=-2)
+    if p == 0:
+        loss = ((x - y) ** p).abs().sum(axis=-1)
+    elif p == float('inf'):
+        loss = paddle.max((x - y).abs(), axis=-1)
+    elif p == float('-inf'):
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py]: p 的取值应该是[0, inf]吧

File: python/paddle/tensor/linalg.py [hunk 5]
```diff
@@ -3350,3 +3350,58 @@ def corrcoef(x, rowvar=True, name=None):
...
+        raise ValueError('The Input(x) and Input(y) should be 2-D tensor.')
+    if x_shape[:-2] != y_shape[:-2] or x_shape[-1] != y_shape[-1]:
+        raise ValueError(
+            "Input(x) shape[:-2] should be same as Input(y) shape[:-2] "
+            "and Input(x) shape[-1] shoule be same as Input(y) shape[-1]"
+        )
+    y = paddle.concat([y] * x_shape[-2], axis=-2)
+    x = paddle.repeat_interleave(x, y_shape[-2], axis=-2)
+    if p == 0:
+        loss = ((x - y) ** p).abs().sum(axis=-1)
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py]: loss 的命名有些不妥

File: python/paddle/fluid/tests/unittests/test_cdist.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            b_i = np.random.random((2, 3, 1, 5)).astype(self.data_type)
+            a = paddle.to_tensor(a_i)
+            b = paddle.to_tensor(b_i)
+            c = paddle.cdist(a, b, p)
+            np.testing.assert_allclose(
+                cdist(a_i, b_i, p), c.numpy(), rtol=1e-05
+            )
+        paddle.enable_static()
+
+    def test_grad_x(self):
```
Comment (GGBond8488) [inline python/paddle/fluid/tests/unittests/test_cdist.py]: 这个命名含义不清楚

File: python/paddle/fluid/tests/unittests/test_cdist.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            )
+            exe = fluid.Executor(place)
+            out = exe.run(
+                fluid.default_main_program(),
+                feed={'x': x_i, 'y': y_i},
+                fetch_list=[result],
+            )
+            np.testing.assert_allclose(cdist(x_i, y_i, p), out[0], rtol=1e-05)
+
+    def test_order(self):
```
Comment (GGBond8488) [inline python/paddle/fluid/tests/unittests/test_cdist.py]: 命名改一下吧

File: python/paddle/fluid/tests/unittests/test_cdist.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        loss = np.max(np.abs(new_x - new_y), axis=-1)
+    elif p == float('-inf'):
+        loss = np.min(np.abs(new_x - new_y), axis=-1)
+    else:
+        loss = np.sum(np.abs((new_x - new_y) ** p), axis=-1) ** (1 / p)
+    loss = loss.reshape((*x.shape[:-2], x.shape[-2], y.shape[-2]))
+    return loss
+
+
+class TestDistAPI(unittest.TestCase):
```
Comment (GGBond8488) [inline python/paddle/fluid/tests/unittests/test_cdist.py:46]: 测试测试不同shape的case

File: python/paddle/tensor/linalg.py [hunk 9]
```diff
@@ -3350,3 +3350,58 @@ def corrcoef(x, rowvar=True, name=None):
         c = paddle.clip(c, -1, 1)
 
     return c
+
+
+def cdist(x, y, p=2, name=None):
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py]: 附上rfc设计文档链接

File: python/paddle/tensor/linalg.py [hunk 10]
```diff
@@ -3350,3 +3350,58 @@ def corrcoef(x, rowvar=True, name=None):
...
+    x_shape = x.shape
+    y_shape = y.shape
+    if len(x_shape) < 2 or len(y_shape) < 2:
+        raise ValueError('The Input(x) and Input(y) should be 2-D tensor.')
+    if x_shape[:-2] != y_shape[:-2] or x_shape[-1] != y_shape[-1]:
+        raise ValueError(
+            "Input(x) shape[:-2] should be same as Input(y) shape[:-2] "
+            "and Input(x) shape[-1] shoule be same as Input(y) shape[-1]"
+        )
+    y = paddle.concat([y] * x_shape[-2], axis=-2)
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py]: 这里的concat 以及 repeat_interleave 是相同的作用？

File: python/paddle/tensor/linalg.py (hunk 2 above)
Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 改为 should be at least 2-D tensor

File: python/paddle/tensor/linalg.py (hunk 1 above)
Comment (yangguohao) [reply python/paddle/tensor/linalg.py:3386]: 添加了检查

File: python/paddle/tensor/linalg.py (hunk 3 above)
Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 是可以不相等的，但是每一个维度需要相等，增加了每一个维度的检查。

File: python/paddle/tensor/linalg.py (hunk 4 above)
Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 已修改，去掉 -inf 的情况

File: python/paddle/tensor/linalg.py (hunk 5 above)
Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 已修改为out

File: python/paddle/fluid/tests/unittests/test_cdist.py (hunk 6 above)
Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_cdist.py]: 已修改

File: python/paddle/fluid/tests/unittests/test_cdist.py (hunk 7 above)
Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_cdist.py]: 已修改

File: python/paddle/fluid/tests/unittests/test_cdist.py (hunk 8 above)
Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_cdist.py:46]: 增加不同 case

File: python/paddle/tensor/linalg.py (hunk 10 above)
Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 这里作用不同，例如 x 为[[2], [3]]， y为[[4],[5]]，计算的时候应该是 2 和4，5分别计算一次距离， 3和4，5分别计算一次距离，这时候就需要 x 为[[2],[2],[3],[3]] y扩展为[[4],[5],[4],[5]]来进行分别计算

Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 或者可能有其他的方法可以做到

File: python/paddle/tensor/linalg.py (hunk 9 above)
Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 这里我还没有提交 rfc，我现在提交等着合入后附上链接。

Comment (GGBond8488) [reply python/paddle/tensor/linalg.py]: cdist应该已有设计文档了吧，api命名以及参数都需要和rfc文档保持一致

Comment (yangguohao) [reply python/paddle/tensor/linalg.py]: 原来的 cdist 设计文档不是通过在 python 端组合的形式

File: python/paddle/tensor/linalg.py [hunk 11]
```diff
@@ -3350,3 +3350,109 @@ def corrcoef(x, rowvar=True, name=None):
...
+
+    for i in range(len(x_shape[:-2])):
+        if x_shape[i] != y_shape[i] and (x_shape[i] != 1 and y_shape[i] != 1):
+            raise ValueError(
+                f"The shape of Input(x) should be same as Input(y) or be broadcast at dimension {i}"
+                f"but got Input(x) {x_shape[i]} and Input(y) {y_shape[i]}"
+            )
+
+    p = float(p)
+
```
Comment (GGBond8488) [inline python/paddle/tensor/linalg.py:3438]: 需要考虑某一维度有0的场景

File: python/paddle/fluid/tests/unittests/test_cdist.py (hunk 8 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/test_cdist.py:46]: 重新写一下单测类，把共同的测试写到父类，不同的测试case写到子类。然后需要单独增加不同compute_mode计算结果是否一致的测试

**Comment by luotao1** [comment]
https://github.com/PaddlePaddle/Paddle/pull/53836#pullrequestreview-1450189767
赛题2已经技术锁定了
